HPS/Heartland/Model/Payment.php

```diff
@@ -368,129 +275,629 @@ public function refund(\Magento\Payment\Model\InfoInterface $payment, $amount)
+                                                       $canSaveToken);
+
+                    $payment->setParentTransactionId($response->transactionId . '-' . $this->transactionTypeMap[ $paymentAction ]);
+                    break;
+                /**
+                 * Reverses the full amount and removes any related capture from the batch*/
+                case (\HpsTransactionType::VOID): // Portico CreditVoid \HpsTransactionType::VOID
+                    $response = $chargeService->void($parentPaymentID);
+                    break;
+                case (\HpsTransactionType::REVERSE):// Portico CreditReversal \HpsTransactionType::REVERSE
+                    $newAuthAmount = $reportTxnDetail->settlementAmount - $requestedAmount;
+                    $response      = $chargeService->reverse($parentPaymentID,
+                                                             $reportTxnDetail->authorizedAmount,
+                                                             $currency,
+                                                             $details,
+                                                             $newAuthAmount);
+                    break;
+                case (\HpsTransactionType::REFUND):// Portico CreditReturn \HpsTransactionType::REFUND
+                    $response = $chargeService->refund($requestedAmount,
+                                                       $currency,
+                                                       $parentPaymentID,
+                                                       $validCardHolder,
+                                                       $details);
+                    break;
+                default:
+                    throw new LocalizedException(new Phrase(__($paymentAction . ' not implemented')));
+            }
+            // even if the MUPT save fails the transaction should still complete so we execute this step first
+
+            /**
+             * @var \Magento\Payment\Model\InfoInterface|\Magento\Payment\Model\Method\AbstractMethod|\Magento\Framework\DataObject $info
+             * @method string $info::getCcNumber() Retrieves the value in the Credit card field in this instance
+             * @method string getCcType() Retrieves the text type for the credit card in this instance
+             */
+            $info = $this->getInfoInstance();
+            $CcL4 = $info->getCcNumber();;
+
+            $this->log($response, 'setStatus ');
+            // set items always found in the response header
+            /** @var \HpsTransaction $response Properties found in the header */
+            //$payment->setStatus($response->responseText);
+            $payment->setTransactionId($response->transactionId . '-' . $this->transactionTypeMap[ $paymentAction ]);
+            $payment->setAdditionalInformation(serialize($response));
+            if ($payment->isCaptureFinal($requestedAmount)) {
+                $payment->setShouldCloseParentTransaction(false);
```

Comment:
method isCaptureFinal will check whether the whole amount is captured. In that case the transaction needs to be closed